### PR TITLE
Fix unsafe creation of temp file for caching, and improve exception raising

### PR DESCRIPTION
**Pull request by kvanzuijlen**
Changes proposed in this pull request:

 * Fixed unsafe creation of temp file for caching
 * Improved the way how exceptions are raised. Exceptions explicitly raised in the code are now (subclasses of) `PyLastError`.

### CODE REVIEW

File: src/pylast/__init__.py [hunk 1]
```diff
@@ -783,7 +785,7 @@ class _ShelfCacheBackend:
     """Used as a backend for caching cacheable requests."""
 
     def __init__(self, file_path=None):
-        self.shelf = shelve.open(file_path)
+        self.shelf = shelve.open(file_path, flag="n")
```
Comment (hugovk) [inline src/pylast/__init__.py:788]: Value | Meaning
-- | --
'r' | Open existing database for reading only (default)
'w' | Open existing database for reading and writing
'c' | Open database for reading and writing, creating it if it doesn’t exist
'n' | Always create a new, empty database, open for reading and writing

The default is `"c"`.

I think we'd want the cache to persist between sessions, for people who run one-off scripts.

Comment (kvanzuijlen) [reply src/pylast/__init__.py]: Didn't think that one through, should be fixed now though. It now always creates a new one if no file path was provided and opens it if it was provided (creating it if it doesn't exist).

Another solution would be to create a `shelve.Shelf`, which would create an in-memory shelf. This way, no temporary files will have to be created and if someone wants a persistent cache, they can pass a file path. What are your thoughts on this?

**Comment by djmattyg007** [comment]
What's preventing this PR from being merged? The exception hierarchy will be most welcome for consumers of this library.

**Comment by hugovk** [comment]
Thanks for the ping, and thanks for the PR!

**Comment by djmattyg007** [comment]
Thanks @kvanzuijlen and @hugovk!

**Pull request merged**